### PR TITLE
Fix for #1111

### PR DESCRIPTION
**Pull request by dmitry-lipetsk**
Hello,

It is a fix for #1111.

There are added a pair of new functions:
- find_exist_or_add_new_global_user
- find_exist_or_add_new_global_credentials

They replaces calls of:
- find_global_user + add_global_user
- find_global_credentials + add_global_credentials

We checks the result of these new functions to detect OOM.

Also, these functions will be a little fast, because old add_xxx functions call find_xxx again.

Note: add_global_credentials is not used in code and can be removed, I can do it if you want.

### CODE REVIEW

File: include/objects.h [hunk 1]
```diff
@@ -71,6 +71,9 @@ bool pop_outstanding_request(PgSocket *client, char *types, bool *skip);
 bool clear_outstanding_requests_until(PgSocket *server, char *types) _MUSTCHECK;
 bool queue_fake_response(PgSocket *client, char request_type) _MUSTCHECK;
 
+PgGlobalUser * find_exist_or_add_new_global_user(const char *name, const char *passwd) _MUSTCHECK;
+PgCredentials * find_exist_or_add_new_global_credentials(const char *name, const char *passwd) _MUSTCHECK;
```
Comment (JelteF) [inline include/objects.h:75]: nit let's name these functions as follows:

```suggestion
PgGlobalUser * find_or_add_new_global_user(const char *name, const char *passwd) _MUSTCHECK;
PgCredentials * find_or_add_new_global_credentials(const char *name, const char *passwd) _MUSTCHECK;
```

File: src/admin.c [hunk 2]
```diff
@@ -1746,13 +1746,10 @@ void admin_setup(void)
 		die("cannot create admin pool?");
 	admin_pool = pool;
 
-	/* user */
-	user = find_global_user("pgbouncer");
+	/* find an exist user or create a new fake user with disabled psw */
```
Comment (JelteF) [inline src/admin.c:1749]:
```suggestion
	/* find an existing user or create a new fake user with disabled password */
```

File: src/client.c [hunk 3]
```diff
@@ -485,9 +485,7 @@ bool set_pool(PgSocket *client, const char *dbname, const char *username, const
 			 * see if the global auth_user is set and use that.
 			 */
 			if (!client->db->auth_user_credentials && cf_auth_user) {
-				client->db->auth_user_credentials = find_global_credentials(cf_auth_user);
-				if (!client->db->auth_user_credentials)
-					client->db->auth_user_credentials = add_global_credentials(cf_auth_user, "");
+				client->db->auth_user_credentials = find_exist_or_add_new_global_credentials(cf_auth_user, "");
```
Comment (JelteF) [inline src/client.c:488]: This should fail too right if it's NULL?

File: src/objects.c [hunk 4]
```diff
@@ -488,33 +488,48 @@ PgDatabase *register_auto_database(const char *name)
...
+		return create_new_global_user(name, passwd);
 
-	passwd = passwd ? passwd : "";
-	safe_strcpy(user->credentials.passwd, passwd, sizeof(user->credentials.passwd));
-	user->credentials.dynamic_passwd = strlen(passwd) == 0;
-	return user;
+	return update_global_user_passwd(user, passwd);
 }
 
+//TODO: is not used.
```
Comment (JelteF) [inline src/objects.c:532]: let's remove the function then.

File: src/objects.c [hunk 5]
```diff
@@ -1002,6 +1022,28 @@ bool queue_fake_response(PgSocket *client, char request_type)
 	return res;
 }
 
+/* Find an exist global user or add a new global user */
```
Comment (JelteF) [inline src/objects.c:1025]:
```suggestion
/* Find an existing global user or add a new global user */
```

File: src/objects.c [hunk 6]
```diff
@@ -1002,6 +1022,28 @@ bool queue_fake_response(PgSocket *client, char request_type)
...
+{
+	PgGlobalUser *user = find_global_user(name);
+
+	if (!user)
+		user = create_new_global_user(name, passwd);
+
+	return user;
+}
+
+/* Find an exist global credentials or add a new global credentials */
```
Comment (JelteF) [inline src/objects.c:1036]:
```suggestion
/* Find an existing global credentials or add a new global credentials */
```

**Review by JelteF — CHANGES_REQUESTED** [review]

**Comment by JelteF** [comment]
Thank you for the contribution. I left some feedback on the code.

File: src/client.c (hunk 3 above)
Comment (dmitry-lipetsk) [reply src/client.c:488]: Should I write something like this

```c
slog_error(client, "set_pool(): failed to allocate a new global credentials");
disconnect_client(client, true, "bouncer resources exhaustion");
return false;
```
?

I will rename these new functoin and will return to this problem.

File: include/objects.h (hunk 1 above)
Comment (dmitry-lipetsk) [reply include/objects.h]: Done. Thanks!

**Comment by dmitry-lipetsk** [comment]
Hello,

I hope, all the requested changes were done. Or no?

File: src/objects.c [hunk 7]
```diff
@@ -496,39 +496,45 @@ PgDatabase *register_auto_database(const char *name)
...
+
+	aatree_insert(&user_tree, (uintptr_t)user->credentials.name, &user->credentials.tree_node);
+	user->pool_mode = POOL_INHERIT;
+	user->pool_size = -1;
+
+	return update_global_user_passwd(user, passwd);
+}
+
+/* add or update client users */
+PgGlobalUser *add_global_user(const char *name, const char *passwd)
```
Comment (JelteF) [inline src/objects.c:530]: This does basically the same as `find_or_add_global_user`, except that it also updates the password of the global user. 

Let's move these functions closer together and make this difference clear from the naming and function comment.

Comment (JelteF) [reply src/objects.c:530]: Actually, maybe it's better to completely remove `add_global_user` and explicitely call `update_global_user_passwd` in all places where we need to update the password.

**Comment by JelteF** [comment]
@dmitry-lipetsk are you still planning to move this forward? 

On a related note, there's now a discord channel to discuss PgBouncer development, feel free to join there: https://discordapp.com/channels/1258108670710124574/1300532992304742481

**Comment by dmitry-lipetsk** [comment]
> @dmitry-lipetsk are you still planning to move this forward?

Hello @JelteF,

I am sorry but I completely lost the context of this project and do not have free time to get it back :-)

You can do with this PR all what you want.

> On a related note, there's now a discord channel to discuss PgBouncer development, feel free to join there: https://discordapp.com/channels/1258108670710124574/1300532992304742481

Thanks but, it seems to me, this resource is not available in my country.

**Pull request closed without merging**